### PR TITLE
bpo-37970: update and improve urlparse and urlsplit doc-strings

### PR DESCRIPTION
**Pull request by idomic**
https://bugs.python.org/issue37970

### CODE REVIEW

**Comment by idomic** [comment]
Hey @zware, would love to get feedback and close this PR.

Ido

**Comment by zware** [comment]
Hi @idomic, sorry I haven't had time to get back to this yet.  Some quick high-level feedback, though:

- CI is failing, so that will need to be addressed.  Without digging in and looking, the most likely culprit is trailing whitespace; run `make patchcheck` if you're on a *nix platform, or run the same commands that does if you're on Windows (yes, terrible UX there, sorry :( ).
- There should be a NEWS entry for this, see the `Details` link on the `bedevere/news` check.
- `urlparse` needs approximately the same treatment.
- It would be good to go through the rest of the module and see if there are other similarly insufficient docstrings that could be updated while we're in there.

**Comment by idomic** [comment]
@zware fixed all of the above 👍

File: Lib/urllib/parse.py
Comment (zware) [inline Lib/urllib/parse.py:451]: Please have a look at PEP 257 for docstring formatting guidelines.

File: Misc/NEWS.d/next/Library/2019-12-03-17-38-02.bpo-37970.f7Vrwm.rst
Comment (zware) [inline Misc/NEWS.d/next/Library/2019-12-03-17-38-02.bpo-37970.f7Vrwm.rst]:
```
Updated :func:`urllib.parse.urlsplit` docstring.
```

**Review by zware — CHANGES_REQUESTED** [review]
Still need to update `urlparse` as well.

**Comment by csabella** [comment]
@idomic, please address the review comments.  Thanks!

**Comment by idomic** [comment]
Thanks @zware @csabella and sorry for the delay!
Fixed the fixed the review comments, let's see if the CI passes.

**Comment by idomic** [comment]
> Thanks @zware @csabella and sorry for the delay!
> Fixed the fixed the review comments, let's see if the CI passes.

@zware @csabella looks like everything passed fine?

**Comment by idomic** [comment]
Hey @taleinat adding you to accelerate this PR

**Comment by taleinat** [comment]
Hi @idomic,

It seems you've rebased your branch on a later commit on the master branch, and then merged that with the existing PR branch when pushing it.

For future reference, for PR branches, I highly recommend only merging rather than rebasing. If you have rebased, though, you should force-push the rebased branch (i.e. `git push -f origin <branch-name>`).

In this case, I suggest you (hard-) reset your branch to the commit before the merge, make sure it's okay, and then force-push that.

**Comment by idomic** [comment]
Thanks @taleinat Done, let's see if the CI passes

**Comment by taleinat** [comment]
@idomic, looks good now!

For future reference: Please leave a comment with the words `I have made the requested changes; please review again` after a committer has requested changes; this will trigger a bot to update the state tags and notify the reviewer.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: This is implied directly by the previous sentence, since this is the major feature of named tuples. IMO it would be better to omit this extra sentence. (Same goes for `urlsplit()` below.)

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: Copy/paste error here: This returns a `ParseResult` object.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: I find this first line less clear than the original (though it is technically correct.)

How about:
```python
"""Parse a URL into 6 components:
<scheme>://<netloc>/<path>;<params>?<query>#<fragment>

The result is a ParseResult object, which is a named 6-tuple with
fields corresponding to the above.

...
```

This then followed by the later description of the other input parameters and the final note regarding % escapes not being expanded.

(The same goes for `urlsplit()` as well.)

**Review by taleinat — CHANGES_REQUESTED** [review]
As mentioned by @zware, the suggested doc-strings don't follow the guidelines in [PEP 257](https://www.python.org/dev/peps/pep-0257/). In addition to my other comments, please fix their formatting.

File: Misc/NEWS.d/next/Documentation/2020-02-11-13-22-06.bpo-379070.aaZnkk.rst
Comment (taleinat) [inline Misc/NEWS.d/next/Documentation/2020-02-11-13-22-06.bpo-379070.aaZnkk.rst]: This PR changes the doc-strings, not "the documentation", which is what goes up to docs.python.org.

Also, IMO this doesn't warrant a NEWS entry, so it can and should be removed. I'm adding the `skip news` label to allow that.

**Comment by idomic** [comment]
@taleinat Thanks for all the feedback, I noticed I forgot the space line after the doc-string summary.
Did took your version as it clarifies the role in less lines.

**Comment by idomic** [comment]
I have made the requested changes; please review again

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: This will actually be a `ParseResult` or a `ParseResultBytes` object, depending on the type of the `url` parameter. It will never be a `SplitResult` object.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: This will actually be a `SplitResult` or a `SplitResultBytes` object, depending on the type of the `url` parameter.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: Please add another empty line here to separate the paragraphs.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: Please add another empty line here to separate the paragraphs.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py:384]: Please add another empty line here to separate the paragraphs.

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py:449]: Please add another empty line here to separate the paragraphs.

**Review by taleinat — CHANGES_REQUESTED** [review]

**Comment by idomic** [comment]
I have made the requested changes; please review again

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: The parameter is named "url" rather than "URL", let's keep the correct spelling.
```suggestion
    depending on the type of the url parameter.
```

File: Lib/urllib/parse.py
Comment (taleinat) [inline Lib/urllib/parse.py]: How about, rather than this and the previous paragraphs being separate, something like:

"The result is a named 6-tuple with fields corresponding to the above. It is either a ParseResult or a ParseResultBytes object, depending on the type of the URL parameter."

File: Misc/NEWS.d/next/Documentation/2020-02-11-13-22-06.bpo-379070.aaZnkk.rst
Comment (taleinat) [reply Misc/NEWS.d/next/Documentation/2020-02-11-13-22-06.bpo-379070.aaZnkk.rst]: @idomic, please remove this file.

**Review by taleinat — CHANGES_REQUESTED** [review]
Almost done! Two more comments, relevant to both doc-strings.

(Also please remove the NEWS entry file.)

File: Lib/urllib/parse.py
Comment (idomic) [reply Lib/urllib/parse.py]: It's shorter, changed to this version

**Comment by idomic** [comment]
I have made the requested changes; please review again

**Comment by idomic** [comment]
@taleinat Should I just cherry pick and create a new PR? I see some of the ci fails even after pulling from the master latest version

**Comment by taleinat** [comment]
@idomic, please don't create a new PR, and there's no need. What you should do is:

1. Hard-reset your branch to the latest master branch
2. Make the latest appropriate changes (even copy-paste from GitHub works)
3. Review, making sure everything is fine
4. run `make patchcheck` (or run `patchcheck.py` directly on Windows)
5. Create a new commit with appropriate prefix: `bpo-37970:`
6. Force-push into this PR's branch

**Comment by idomic** [comment]
I have made the requested changes; please review again

**Comment by taleinat** [comment]
@idomic, it seems something went very wrong on your end with this PR's branch. I've gone ahead and force-pushed a fixed version, with a single commit.

If you want to check it out locally, note that you'll need to delete your local branch (or rename it to back it up) before pulling this branch.

**Review by taleinat — APPROVED** [review]

**Comment by taleinat** [comment]
Many thanks for the PR, @idomic, and for your patience repeatedly waiting for a review and making the requested changes!

I assure you that for changes such as this, it usually doesn't take this long nor as many iterations as it took us here. I hope this experience doesn't discourage you from continuing to contribute to the project!

**Comment by idomic** [comment]
@taleinat Not at all, I'm happy to learn and help!
Hopefully our next work will be more concise :)

**Pull request merged**
